src/hotspot/share/code/nmethod.cpp

```diff
@@ -2139,7 +2286,15 @@ void nmethod::purge(bool unregister_nmethod) {
   delete[] _compiled_ic_data;
 
   if (_immutable_data != blob_end()) {
-    os::free(_immutable_data);
+    long _immutable_data_references = *immutable_data_references_begin();
```

Comment:
Why `long`? It should be int for all platforms (we still have 32-bit arm).
Check the current value and update it should be done by nmethod's inline functions.